### ISSUE TITLE
New FHS for Nitrux

### ISSUE THREAD

**Issue by UriHerrera** [Nitrux/nitrux-iso-tool#35]
reserved.

**Comment by UriHerrera** [comment]
Topics of interest:

* From the creator of AppImage https://github.com/probonopd/hello/wiki

> Desktop operating system with a focus on simplicity and usability (think Macintosh System 1).
> Based on the Linux kernel but without the complexities of normal Linux distributions.

**Comment by UriHerrera** [comment]
### Define Directory structure

To move forward, we need to organize how the structure will be; this is an example.

```
 /boot/ -> /Core/Boot
 /cdrom/ -> /Core/Boot/ESP
 /dev/ -> /Devices
 /etc/ -> /System/Configuration
 /home/ -> /Users
 /isodevice/ -> /Core/System/Deployments
 /media/ -> /Devices/Removable
 /mnt/ -> /System/Mount/Filesystems
 /opt/ -> /System/Binaries/Optional
 /proc/ -> /System/Processes
 /run/ -> /System/Runtime
 /srv/ -> /System/Server/Services
 /sys/ -> /System/DevicesFS
 /tmp/ -> /System/TempFS
 /usr/bin/ -> /System/Binaries
 /usr/lib/ -> /System/Libraries
 /usr/share/ -> /System/Resources/Shared
 /var/ -> /System/Variable
```

Most would be under `/System`, and as per the original idea, the names are descriptive of their contents.

**Comment by UriHerrera** [comment]
### Progress

This is a _partial_ implementation, of the above structure. Works as intended, but there's the side effect that the sidebar of Index is cluttered.

![VirtualBox_Nitrux_13_05_2020_07_09_57](https://user-images.githubusercontent.com/3053525/81810943-776acc00-94e9-11ea-82dc-c56f04c6e985.png)
![VirtualBox_Nitrux_13_05_2020_07_10_10](https://user-images.githubusercontent.com/3053525/81810944-776acc00-94e9-11ea-9550-9763bec58566.png)
![VirtualBox_Nitrux_13_05_2020_07_10_22](https://user-images.githubusercontent.com/3053525/81810947-78036280-94e9-11ea-84f0-48422d92f23a.png)
![VirtualBox_Nitrux_13_05_2020_07_11_03](https://user-images.githubusercontent.com/3053525/81810948-78036280-94e9-11ea-9c59-d3e6406976b6.png)
![VirtualBox_Nitrux_13_05_2020_07_11_06](https://user-images.githubusercontent.com/3053525/81810950-789bf900-94e9-11ea-8c91-30b00c15eaf5.png)
![VirtualBox_Nitrux_13_05_2020_07_11_14](https://user-images.githubusercontent.com/3053525/81810952-789bf900-94e9-11ea-940c-c3c0ff72822e.png)
![VirtualBox_Nitrux_13_05_2020_07_11_22](https://user-images.githubusercontent.com/3053525/81810953-789bf900-94e9-11ea-9359-2e94adcf578c.png)
![VirtualBox_Nitrux_13_05_2020_07_11_28](https://user-images.githubusercontent.com/3053525/81810955-79348f80-94e9-11ea-8149-cd08aacf9051.png)

**Comment by luis-lavaire** [comment]
It seems you're mounting those as overlays. That new structure should be built based on links, not mount points.

**Comment by luis-lavaire** [comment]
BTW, do we _really_ want to expose all those system interfaces to the users?

> ```
>  /boot/ -> /Core/Boot
>  /cdrom/ -> /Core/Boot/ESP
>  /dev/ -> /Devices
>  /etc/ -> /System/Configuration
>  /home/ -> /Users
>  /isodevice/ -> /Core/System/Deployments
>  /media/ -> /Devices/Removable
>  /mnt/ -> /System/Mount/Filesystems
>  /opt/ -> /System/Binaries/Optional
>  /proc/ -> /System/Processes
>  /run/ -> /System/Runtime
>  /srv/ -> /System/Server/Services
>  /sys/ -> /System/DevicesFS
>  /tmp/ -> /System/TempFS
>  /usr/bin/ -> /System/Binaries
>  /usr/lib/ -> /System/Libraries
>  /usr/share/ -> /System/Resources/Shared
>  /var/ -> /System/Variable
> ```

Because it's not that the original files and directories will not exist, they will just be hidden. I think that the point is not just to rename stuff while also keeping everything, but to actually hide most of the underlying system from the user (thus, decluttering the view of the filesystem). I'd prefer that most of the `/System` directory wasn't present, preserving only:

```
/etc		->  /System/Configuration
/media		->  /System/Devices
/isodevice	->  /System/Deployments
/home		->  /Users
/opt		->  /Applications
```

An end-user ─as per our archetypal idea of end-user─ should not deal with processes, filesystems, pseudo-devices, program runtime data, temporal data, boot files, etc.

**Comment by UriHerrera** [comment]
> It seems you're mounting those as overlays. That new structure should be built based on links, not mount points.

No, they're not overlays, they are mount binds. KIO displays them as overlays. Index **should** hide them.

> BTW, do we _really_ want to expose all those system interfaces to the users?

Yes, the intention wasn't just to hide the Linux FHS, but to _use_ a new FHS, this is but a stopgap to get there. As it turns out, the FHS in macOS is visually similar to this, but as it is, Finder hides what they call _the legacy structure_ from the user by default, and doesn't list them in its sidebar.

> Because it's not that the original files and directories will not exist, they will just be hidden. I think that the point is not just to rename stuff while also keeping everything, but to actually hide most of the underlying system from the user (thus, decluttering the view of the filesystem). I'd prefer that most of the /System directory wasn't present, preserving only:
> 
> /etc		->  /System/Configuration
> /media		->  /System/Devices
> /isodevice	->  /System/Deployments
> /home		->  /Users
> /opt		->  /Applications
>
> An end-user ─as per our archetypal idea of end-user─ should not deal with processes, filesystems, pseudo-devices, program runtime data, temporal data, boot files, etc.

Indeed, it doesn't solve it; it's a partial implementation of the general idea. The point isn't either to lockout users from their operating system or viewing the internals of their operating system; _you shouldn't be prevented from looking under the hood of your car to fix it_.

By using `/System`, we let the user know that everything under here will affect their operating system. Any change done here can and will have a consequence, and the same with the other directories.

So the next step should appropriately hide the UNIX structure leaving only the new one, which is more readable.

**Comment by luis-lavaire** [comment]
> Yes, the intention isn't just to hide the Linux FHS, but to _use_ a new FHS...

That's unfeasible for us: We would need to patch almost every program that expects any of those locations (many programs do; many). That's why I see this more as a simple cosmetical concern, not a technical one (it's perceived in this way by the GoboLinux developers too).

> We can also find another way to prevent non-admin users from accessing critical directories like `/System` using methods like jailing if we wanted that, though in that case, it should be an option, not a default.

Users, groups, and file permissions _**already handle this**_! We should not reinvent the wheel. 

> Indeed, it doesn't solve it; it's a partial implementation of the general idea. The point isn't either to lockout users from their operating system or viewing the internals of their operating system; _you shouldn't be prevented from looking under the hood of your car to fix it_.

Keeping the old structure, with the only difference of it being hidden, will save us a **LOT** of work, will ensure backward compatibility with any software running in the system, and also, _"let you look under the hood"_ if you want to.

My advice is that we shouldn't see this as technical improvement, but an aesthetic one.

**Comment by UriHerrera** [comment]
> That's unfeasible for us: We would need to patch almost every program that expects any of those locations (many programs do; many). That's why I see this more as a simple cosmetical concern, not a technical one (it's perceived in this way by the GoboLinux developers too).

No, that's not what I'm saying. Of course, we can't patch everything. The point is, we don't need to the _legacy_ FHS paths (which we don't), so we don't need to show them (so we _hide them_, not _replace_ them); instead we show the new ones (which are a cosmetic change, right now); therefore, the new ones are the ones that are _used_.

> Users, groups, and file permissions already handle this! We should not reinvent the wheel.

I also didn't say that we would want to. What I'm saying is, if we didn't want users poking around the critical directories, there are other ways to achieve that other than not displaying the directories at all, because that wouldn't be very open.

> Keeping the old structure, with the only difference of it being hidden, will save us a LOT of work

_**Yes, that is what I'm saying**_.

**Comment by UriHerrera** [comment]
# Update

So, it's almost four years since we last checked this. 

TL;DR 

I've reimplemented the FHS in what would be the next release. It's visible in both the Live session, as when we used *znx*, and now visible in an installed system, as we use Calamares.

Also, it's visible using Index or other graphical file managers and command-line commands like `ls` and `tree`, using the magic of not patching stuff.

It's still not _exactly_ what I proposed, but it's good enough for me.

Here are some screenshots.

![VirtualBox_Nitrux VM_12_11_2023_00_47_36](https://github.com/Nitrux/iso-tool/assets/3053525/8c96fc2b-4d3a-462b-a85c-f3d5e9734ed0)
> Using `ll` in the Live session.

![VirtualBox_Nitrux VM_12_11_2023_00_47_48](https://github.com/Nitrux/iso-tool/assets/3053525/f96a22e7-4301-4173-9e7b-9a1c54d383cb)
> Using `tree ' in the Live session.

![VirtualBox_Nitrux VM_12_11_2023_00_48_02](https://github.com/Nitrux/iso-tool/assets/3053525/24f76cdc-acfe-4e9c-8071-21cb76e53439)
> Using Index in the Live session.

---

I'll move this to a discussion because we can still take it further.
